lightning/src/util/test_utils.rs

```diff
@@ -795,6 +797,12 @@ impl NodeSigner for TestNodeSigner {
 		unreachable!()
 	}
 
+	fn sign_bolt12_message(
+		&self, _message: &TaggedBytes, _metadata: &[u8]
```

Comment:
@devrandom @ksedgwic any feedback on this API?

Reply:
isn't the metadata also tagged?  would be good in general to have more specific comments about the format

Reply:
BTW, the `_metadata` arg can stay if it provides performance advantages in some applications.  validating signers can just ignore it.

Reply:
This is what we use for CLN:
```
/// SignBolt12
#[derive(SerBolt, Debug, Serialize, Deserialize)]
#[message_id(25)]
pub struct SignBolt12 {
    pub message_name: WireString,
    pub field_name: WireString,
    pub merkle_root: Sha256,
    pub public_tweak: Octets,
}

///
#[derive(SerBolt, Debug, Serialize, Deserialize)]
#[message_id(125)]
pub struct SignBolt12Reply {
    pub signature: Signature,
}
```

Reply:
> isn't the metadata also tagged? would be good in general to have more specific comments about the format

The metadata was included simply to provide a way to derive the signing keys from the metadata. However, after the discussion in https://github.com/lightningdevkit/rust-lightning/pull/1989#discussion_r1149774759, it's unclear if this is actually needed.

As currently implemented, LDK's `ChannelManager` will either [sign the invoice from derived keys](https://github.com/lightningdevkit/rust-lightning/pull/2371/commits/a13f3253b25cbae8db19a5459d0b86dcb72c30c5#diff-645d12c7269d09caab57d2e0c1fd34f672ee259994c86ad4bad7f4da9183671cR7273) for you or [delegate to the `NodeSigner`](https://github.com/lightningdevkit/rust-lightning/pull/2371/commits/a13f3253b25cbae8db19a5459d0b86dcb72c30c5#diff-645d12c7269d09caab57d2e0c1fd34f672ee259994c86ad4bad7f4da9183671cR7294). The latter is only reached when the originating offer [did not include any blinded paths](https://github.com/lightningdevkit/rust-lightning/blob/685f26665594e9b941ae94b00600f2b4a4a533fa/lightning/src/offers/offer.rs#L280) and thus included the actual `node_id` for the `signing_pubkey` in the offer. See: https://github.com/lightningdevkit/rust-lightning/blob/685f26665594e9b941ae94b00600f2b4a4a533fa/lightning/src/offers/offer.rs#L158-L159

> This is what we use for CLN:
> 
> ```
> /// SignBolt12
> #[derive(SerBolt, Debug, Serialize, Deserialize)]
> #[message_id(25)]
> pub struct SignBolt12 {
>     pub message_name: WireString,
>     pub field_name: WireString,
>     pub merkle_root: Sha256,
>     pub public_tweak: Octets,
> }
> 
> ///
> #[derive(SerBolt, Debug, Serialize, Deserialize)]
> #[message_id(125)]
> pub struct SignBolt12Reply {
>     pub signature: Signature,
> }
> ```

Does this mean you are blindly signing the tagged hash of the merkle root without computing the merkle root from the `invoice` or `invoice_request` TLV stream on your own?

Where does the tweak come from? Is it CLN-specific?

Reply:
The bytes are now exposed in `TaggedHash`.